File: 백준/Silver/9012. 괄호/괄호.py
tc = int(input())
for test in range(tc):
    s = input()
    stack = []
    top = -1
    result = ''
    for i in s:
        if i == '(':
            stack.append(i)
            top += 1
        elif i == ')':
            if stack:
                stack.pop()
                top -= 1
            else:
                result = 'NO'
                break

    else:
        if not stack:
            result = 'YES'
        else:
            result = 'NO'
    print(result)

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.sql.SQLOutput;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num = Integer.parseInt(br.readLine());

        for(int i = 0; i<num; i++){
            String word = br.readLine();
            Stack<Character> stack = new Stack<>();
            boolean check = true;
            for(char c : word.toCharArray()) {

                if (c == '(') {
                    stack.push(c);
                } else if(c == ')'){
                    if(stack.isEmpty()){
                        check = false;
                        break;
                    }
                    stack.pop();
                }
            }
            if (check && stack.isEmpty()) {
                System.out.println("YES");
            } else{
                System.out.println("NO");
            }
        }
    }
}
